### ISSUE TITLE
Integrate Socket.IO with NextJS 13 App Router `app/api/chat/route.ts`

### ISSUE THREAD

**Issue by bahmanworld** [socketio/socket.io#4763]
How to integrate Socket.IO into NextJS route.ts?

```typescript
import { Server } from 'socket.io'


export const GET = (request: Request) => {
   // todo
}

export const POST = (request: Request) => {
   // todo
}

...
```

**Comment by Lazhor** [comment]
There are several tutorials on google and YouTube with NextJS, ex:

https://www.showwcase.com/show/18995/real-time-chat-application-under-5-minutes-using-nextjs-and-socketio

**Comment by bahmanworld** [comment]
thanx @Lazhor 
But that's not working with the Nextjs app routing system.
Can you help me a little more?

**Comment by esponges** [comment]
I'm also wondering if this is possible. Did you make any attempts to implement it? @bahmanworld

**Comment by bahmanworld** [comment]
> 



> I'm also wondering if this is possible. Did you make any attempts to implement it? @bahmanworld

I was trying to integrate socketIO once per day last month, nothing happened, i have no idea right now ...

**Comment by mochamadsatria** [comment]
@bahmanworld , @Lazhor  solution works quite well. I mention this because you can actually have an app and page router in a single Next.js project. Vercel sets it up for you, allowing you to adopt it `incrementally`. Maybe you should try it too.  Just make sure the name of the last accessible route is not duplicated between them.

**Comment by bahmanworld** [comment]
Thank you @mochamadsatria for your great advice , but i've already tried both solution .. could you explain more what i miss?

However i never try integrate socketIO with nextjs 14 lately.

**Comment by mochamadsatria** [comment]
@bahmanworld, there is some scenario where your configuration might gone wrong:

1. You put your code inside app router instead of pages router.
2. You use app router but not considering if all component inside it is treated as server component.
3. Your client component configuration inside app router might not correct.

here is the working minimal example, I hope you'll get the point:
https://github.com/mochamadsatria/next14io-example

to enhance it there some trick you can use:

1. Use server action and take advantage of how the server component lifecycle work.
2. Use pages router instead.
3. I know there is a way to make the render-lazy in server component, I have using this method in the past but forgot where the documentation is located XD.

There is a known issue with my example at this moment:
The code work perfectly except it will render the message twice in route inside app router, it will not happen if you use pages router. I have couple solution in mind but I do not have time to try it.

If you can improve the example pls make a pull request into my repository so anyone can see how it works.

**Comment by darrachequesne** [comment]
Hi all!

I've added a guide in the documentation: https://socket.io/how-to/use-with-nextjs

Please ping me if something is not clear enough.

**Comment by bahmanworld** [comment]
⚡️ Now we are talking
Thanks @darrachequesne , It works perfectly 🌺

**Comment by headironc** [comment]
Excuse, how to get the socket instance in server actions

**Comment by nnti3n** [comment]
How to get socket.io server instance in route handler

**Comment by SurajKharkwal** [comment]
@darrachequesne Unfortunately, this solution isn't compatible with Vercel. As mentioned in the Next.js documentation, custom servers cannot be deployed to Vercel. You can find more details here: [Next.js Custom Server Documentation](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server).

**Comment by darrachequesne** [comment]
@SurajKharkwal yes, you are absolutely right, that's because Vercel does not support WebSocket connections.

Reference: https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections

I've added a note at the top of the guide.

@headironc that's a good question! Were you able to find a solution? I guess you should be able to add some kind of cookie, to create a link between the request handlers and Socket.IO.

See also: https://socket.io/how-to/use-with-express-session#using-the-session-id

**Comment by Yokachi** [comment]
@bahmanworld > ⚡️ Now we are talking Thanks @darrachequesne , It works perfectly 🌺

Did u also use nextauth? I tryed the same thing and i keep getting an error aboute my googleProvider is not a function, does someone know why? Thanks!!

**Comment by Yokachi** [comment]
@darrachequesne did u know about this issue? Or is it from my code? Thanks

**Comment by darrachequesne** [comment]
I think this can be closed now. Please reopen if needed!
